### ISSUE TITLE
Tuning Glibc Environment Variables (e.g. MALLOC_ARENA_MAX)

### ISSUE THREAD

**Issue by ipsi** [cloudfoundry/java-buildpack#320]
This is the other issue we discussed @nebhale.

Having been looking into some Java memory issues recently, I came across the following two articles from Heroku (should I just call them "The unnamable ones"? :P):

[Tuning glibc Memory Behavior](https://devcenter.heroku.com/articles/tuning-glibc-memory-behavior#what-value-to-choose-for-malloc_arena_max)
[Testing Cedar-14 Memory Use](https://devcenter.heroku.com/articles/testing-cedar-14-memory-use)

Which suggest that tuning `MALLOC_ARENA_MAX` can help with memory problems in applications. It appears that it also causes problems for [Hadoop](https://issues.apache.org/jira/browse/HADOOP-7154), though that's probably their own fault for monitoring VMEM (seriously, guys, why?!).

In my own testing I have noticed that setting `MALLOC_ARENA_MAX` seems to help slow down memory growth, though it _does not_ prevent it entirely.

There are performance implications from setting `MALLOC_ARENA_MAX` (I've always set it to `2`, as suggested by Heroku), but that falls out like so:
- New thread created, asks for an arena
- If < `MALLOC_ARENA_MAX` blocks in use (defaults to `8 * cpu_count`), acquire one.
- If >, share an existing one
- When `malloc`ing new memory, if have exclusive use of an arena, no need to lock.
- If shared, have to lock, thus resulting in contention for memory. Prior to `glibc` introducing arenas, all threads contended for the same memory.
- When thread exits, `glibc` marks the arena as "free", but does _not_ return it to the OS.

That's my understanding at any rate. So I think most Java apps on CF are basically unaffected for the following reasons:
- Limited amonut of thread-thrashing. If a major concern, can configure container to have `min_threads = max_threads`. Most threads are long-lived and process a lot of requests.
- Java does not do `malloc` on a per-thread basis - it `mallocs` the heap space at start up, and then writes to it as needed. As far as I know, once a thread is running, there should be no need to do an OS-level `malloc`, meaning we'd never have OS-level contention for memory (that's pulled up into the Java runtime).
  - NIO and things like `ByteBuffer`s might result in `malloc` calls - not sure about that.

With all of that said, after digging deeper into this, it seems like setting `MALLOC_ARENA_MAX` may have just hidden a deeper problem, and I'm not sure whether I would advise setting it by default. It didn't seem to adversely affect performance, but both applications I tested were effectively non-performant by design, so don't take that as gospel.

I had initially thought that there may be memory concerns here, but I'm not so sure anymore. I think the following scenarios are what would happen if there were a memory leak:

**Unlimited `MALLOC_ARENAS_MAX`**
- Thread asks for arena
- Thread gets arena
- Thread fills arena, never frees memory
- Thread asks for new arena
- Repeat until `cgroup` OOM killer his.

**`MALLOC_ARENAS_MAX` set to `2`**
- Thread asks for arena
- Thread gets arena
- Thread fills arena, never frees memory
- Thread asks for new arena, but no more available.
- Thread instead writes to "main" arena (or the native program heap in this case), which is unbounded. Given enough time, `cgroup` OOM killer hits.

(Actual allocation decisions are a lot more complex - this is heavily simplified. See [Understanding glibc malloc](https://sploitfun.wordpress.com/2015/02/10/understanding-glibc-malloc/) for an in-depth explanation).

This would explain why I saw the heap space reported by `pmap` (which is _not_ the Java Heap!) going up and up in one of my test runs - there was a memory leak somewhere in native code, the arenas got filled up, and once they were full, it had to fall back to using the native heap.

With all that said, some allowance for the native memory used by individual threads should probably be made. If it's small enough, it could probably stay in the "native" part of the calculator. Otherwise, since it's relative to the number of threads used, it may need to be calculated somehow.

Regardless of the decision to set it, I _do_ think that documenting it would be very useful (along with how it relates to Java, etc), as would providing a link to other [`glibc` configuration environment variables](http://man7.org/linux/man-pages/man3/mallopt.3.html). Finding this information was _not_ as easy as I would have liked. I think this should also be documented for other buildpacks as well - Ruby and Go, particularly, are more likely to see issues with memory arenas - note that the Heroku examples all talked about Ruby.

**Comment by lhotari** [comment]
@ipsi Are you aware of #160 and #163 ?

**Comment by lhotari** [comment]
I just received [a tweet from @TimGerlach from SAP](https://twitter.com/heron_42/status/768721870929604608) and he seems interested in improvements in this area as well. @TimGerlach , Could you elaborate your requirements?

**Comment by lhotari** [comment]
> This would explain why I saw the heap space reported by pmap (which is not the Java Heap!) going up and up in one of my test runs - there was a memory leak somewhere in native code, the arenas got filled up, and once they were full, it had to fall back to using the native heap.

Have you checked [this article written by Evan Jones](http://www.evanjones.ca/java-native-leak-bug.html): 

> TL;DR: Always close GZIPInputStream and GZIPOutputStream since they use native memory via zlib. To track down leaks, use jemalloc and turn on sampling profiling using the MALLOC_CONF environment variable. 

Recently Heroku [blogged about tracking down a similar bug in Kafka](https://blog.heroku.com/fixing-kafka-memory-leak).

**Comment by TimGerlach** [comment]
Thanks @lhotari for mentioning and pointing me to this Issue. I am going to elaborate on #319 as it is closer related to our topic.

**Comment by ipsi** [comment]
@lhotari I'm aware of the first one - was not aware of the second issue. As far as I can tell, there is actually a HotSpot bug, and this might be what you're seeing as well: http://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8164293

I saw the GZIP thing, and even ran my app with JEMalloc, but did not see the same issues that they reported. Given that this issue (a) only appears in Java 8, and (b) disappears when HotSpot is disabled, I suspect that there are multiple issues that could cause apps to increase in memory over time.

**Comment by lhotari** [comment]
@ipsi Thanks for pointing out the JDK bug. 

Did you take a look at the assumptions I presented in https://github.com/cloudfoundry/java-buildpack/issues/163#issue-60842702 ?  
One assumption is that tuning `CodeCacheExpansionSize` and `MinMetaspaceExpansion` would also reduce malloc memory fragmentation. 
There must be multiple factors that cause the same symptom of slow growth of the process RSS. That was the main reason for filing #163 .

**Comment by tootedom** [comment]
Hi there,

I've notice that TieredCompilation is seemingly resulting in continual growth.  Disabling TieredCompilation (-XX:-TieredCompilation), has stopped this growth from occurring on several of our applications (all using jdk8).  I submitting a bug to Oracle today to evaluation, which could be related to the already mentioned bug.  http://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8164293

Just thought I'd mention it, in case this helps you out.

/dom

**Comment by ipsi** [comment]
FYI, the bug I raised has been marked as resolved in https://bugs.openjdk.java.net/browse/JDK-8164293, target version is 8u152, which has an expected release date of 2017-10-16.

**Comment by nebhale** [comment]
Well, that's a hilarious date.

**Comment by mweirauch** [comment]
FYI, the bug has been backported to 8u131 which is released: https://bugs.openjdk.java.net/browse/JDK-8178124

**Comment by nebhale** [comment]
Well that's good news.  I'll leave this open for a couple of weeks to allow people to test against v3.16 and v4.0 and let me know if there is still an outstanding issue.

**Comment by nebhale** [comment]
No complaints, so I'm closing this out.
